urllib3/util/connection.py

```diff
@@ -108,6 +116,134 @@ def _set_socket_options(sock, options):
         sock.setsockopt(*opt)
 
 
+def _connect_happy_eyes(address, timeout=socket._GLOBAL_DEFAULT_TIMEOUT,
+                        source_address=None, socket_options=None):
+    """ Implements the Happy Eyes protocol (RFC 6555) which allows
+    multiple sockets to attempt to connect from different families
+    for better connect times for dual-stack clients where server
+    IPv6 service is advertised but broken. """
+    err = None
+
+    if address in HAPPY_EYES_CACHE:
+        family, proto, expires = HAPPY_EYES_CACHE[address]
+
+        # If the cache entry is expired, don't use it.
+        if time.time() < expires:
+            del HAPPY_EYES_CACHE[address]
+
+        # Otherwise try to use the entry right away, if this
+        # fails then run the Happy Eyes protocol again.
+        else:
+            sock = None
+            try:
+                sock = socket.socket(family, socket.SOCK_STREAM, proto)
+
+                _set_socket_options(sock, socket_options)
+
+                if timeout is not socket._GLOBAL_DEFAULT_TIMEOUT:
+                    sock.settimeout(timeout)
+                if source_address:
+                    sock.bind(source_address)
+
+                sock.connect(address)
+                return sock
+            except socket.error as e:
+                err = e
+                sock.close()
+
+    # Couldn't find a cached value or cached value didn't work.
+    host, port = address
+    socks = []
+
+    # Perform a DNS lookup for the address for IPv4 or IPv6 families.
+    family = allowed_gai_family()
+    for res in socket.getaddrinfo(host, port, family, socket.SOCK_STREAM):
+        af, socktype, proto, canonname, sa = res
+
+        # Create and connect the socket for each address we get back.
+        sock = None
+        try:
+            sock = socket.socket(af, socktype, proto)
+
+            _set_socket_options(sock, socket_options)
+
+            if source_address:
+                sock.bind(source_address)
+            sock.settimeout(0)  # Make the socket non-blocking to use select.
+
+            # We should catch all errors except all non-blocking error codes.
+            errcode = sock.connect_ex(sa)
+            if errcode not in [errno.EINPROGRESS,
+                               errno.EAGAIN,
+                               errno.ENOTCONN]:
+                raise socket.error(errcode)
+
+            socks.append(sock)
+
+        except socket.error as e:
+            err = e
+            if sock is not None:
+                sock.close()
+
+    found = None
+    while socks:
+        select_timeout = timeout
+        if timeout is socket._GLOBAL_DEFAULT_TIMEOUT:
+            select_timeout = None
+
+        # Use select to detect which sockets are connected.
+        _, connected, _ = select([], socks, [], select_timeout)
+
+        # If no sockets are ready, then we timed out.
```

Comment:
Gah, annoying. Ok.

So let me define the constraints of what I want here. I'd like us to have a library that provides an abstraction over the best `select` or `select`-like APIs available on a platform that is suitable both for the HappyEyeballs case and the PyOpenSSL contrib module case. It should handle `EINTR`. It should not bring in a bunch of unimportant code. And it should be maintainable.

If this can't be a straight backport or a third-party dependency, we need to treat it as product code. That means a full suite of tests, 100% test coverage, and all that jazz. We can derive this from selectors if that's a useful starting off point.